README.md
# NAME

Myriad - microservice coördination

[![Coverage status](https://coveralls.io/repos/github/binary-com/perl-Myriad/badge.svg?branch=master)](https://coveralls.io/github/binary-com/perl-Myriad?branch=master)
[![Test status](https://circleci.com/gh/binary-com/perl-Myriad.svg?style=shield&circle-token=55b191c6582ef5932e57b142fb29d8e13ae19598)](https://app.circleci.com/pipelines/github/binary-com/perl-Myriad)
[![Docker](https://img.shields.io/docker/pulls/deriv/myriad.svg)](https://hub.docker.com/r/deriv/myriad)

# SYNOPSIS

    use Myriad;
    Myriad->new->run;

# DESCRIPTION

Myriad provides a framework for dealing with asynchronous, microservice-based code.
It is intended for use in an environment such as Kubernetes to support horizontal
scaling for larger systems.

Overall this framework encourages - but does not enforce - single-responsibility
in each microservice: each service should integrate with at most one external system,
and integration should be kept in separate services from business logic or aggregation.
This is at odds with common microservice frameworks, so perhaps it would be more accurate
to say that this framework is aimed at developing "nanoservices" instead.

## Do you need this?

If you expect to be dealing with more traffic than a single server can handle,
or you have a development team larger than 30-50 or so, this might be of interest.

For a smaller system with a handful of users, it's _probably_ overkill!

# Modules and code layout

- [Myriad::Service](https://metacpan.org/pod/Myriad%3A%3AService) - load this in your own code to turn it into a microservice
- [Myriad::RPC](https://metacpan.org/pod/Myriad%3A%3ARPC) - the RPC abstraction layer, in `$self->rpc`
- [Myriad::Storage](https://metacpan.org/pod/Myriad%3A%3AStorage) - abstraction layer for storage, available as `$self->storage` within services
- [Myriad::Subscription](https://metacpan.org/pod/Myriad%3A%3ASubscription) - the subscription handling layer, in `$self->subscription`

Each of the three abstractions has various implementations. You'd set one on startup
and that would provide functionality through the top-level abstraction layer. Service code
generally shouldn't need to care which implementation is applied. There may however be cases
where transactional behaviour differs between implementations, so there is some basic
functionality planned for checking whether RPC/storage/subscription use the same underlying
mechanism for transactional safety.

## Storage

The [Myriad::Storage](https://metacpan.org/pod/Myriad%3A%3AStorage) abstract API is a good starting point here.

For storage implementations, we have:

- [Myriad::Storage::Redis](https://metacpan.org/pod/Myriad%3A%3AStorage%3A%3ARedis)
- [Myriad::Storage::PostgreSQL](https://metacpan.org/pod/Myriad%3A%3AStorage%3A%3APostgreSQL)
- [Myriad::Storage::Memory](https://metacpan.org/pod/Myriad%3A%3AStorage%3A%3AMemory)

Additional transport mechanisms may be available, see CPAN for details.

## RPC

Simple request/response patterns are handled with the [Myriad::RPC](https://metacpan.org/pod/Myriad%3A%3ARPC) layer ("remote procedure call").

Details on the request are in [Myriad::RPC::Request](https://metacpan.org/pod/Myriad%3A%3ARPC%3A%3ARequest) and the response to be sent back is in [Myriad::RPC::Response](https://metacpan.org/pod/Myriad%3A%3ARPC%3A%3AResponse).

- [Myriad::RPC::Implementation::Redis](https://metacpan.org/pod/Myriad%3A%3ARPC%3A%3AImplementation%3A%3ARedis)
- [Myriad::RPC::Implementation::PostgreSQL](https://metacpan.org/pod/Myriad%3A%3ARPC%3A%3AImplementation%3A%3APostgreSQL)
- [Myriad::RPC::Implementation::Memory](https://metacpan.org/pod/Myriad%3A%3ARPC%3A%3AImplementation%3A%3AMemory)

Additional transport mechanisms may be available, see CPAN for details.

## Subscriptions

The [Myriad::Subscription](https://metacpan.org/pod/Myriad%3A%3ASubscription) abstraction layer defines the available API here.

Subscription implementations include:

- [Myriad::Subscription::Implementation::Redis](https://metacpan.org/pod/Myriad%3A%3ASubscription%3A%3AImplementation%3A%3ARedis)
- [Myriad::Subscription::Implementation::PostgreSQL](https://metacpan.org/pod/Myriad%3A%3ASubscription%3A%3AImplementation%3A%3APostgreSQL)
- [Myriad::Subscription::Implementation::Memory](https://metacpan.org/pod/Myriad%3A%3ASubscription%3A%3AImplementation%3A%3AMemory)

Additional transport mechanisms may be available, see CPAN for details.

## Transports

Note that _some layers don't have implementations for all transports_ - MQ for example does not really provide a concept of "storage".

Each of these implementations is supposed to separate out the logic from the actual transport calls, so there's a separate ::Transport set of classes here:

- [Myriad::Transport::Redis](https://metacpan.org/pod/Myriad%3A%3ATransport%3A%3ARedis)
- [Myriad::Transport::PostgreSQL](https://metacpan.org/pod/Myriad%3A%3ATransport%3A%3APostgreSQL)
- [Myriad::Transport::Memory](https://metacpan.org/pod/Myriad%3A%3ATransport%3A%3AMemory)

which deal with the lower-level interaction with the protocol, connection management and so on. More details on that
can be found in [Myriad::Transport](https://metacpan.org/pod/Myriad%3A%3ATransport) - but it's typically only useful for people working on the [Myriad](https://metacpan.org/pod/Myriad) implementation itself.

## Other classes

Documentation for these classes may also be of use:

- [Myriad::Exception](https://metacpan.org/pod/Myriad%3A%3AException) - generic errors, provides ["throw" in Myriad::Exception](https://metacpan.org/pod/Myriad%3A%3AException#throw) and we recommend that all service errors implement this rôle
- [Myriad::Plugin](https://metacpan.org/pod/Myriad%3A%3APlugin) - adds specific functionality to services
- [Myriad::Bootstrap](https://metacpan.org/pod/Myriad%3A%3ABootstrap) - startup used in `myriad.pl` for providing autorestart and other functionality
- [Myriad::Service](https://metacpan.org/pod/Myriad%3A%3AService) - base class for a service
- [Myriad::Registry](https://metacpan.org/pod/Myriad%3A%3ARegistry) - support for registering services and methods within the current process
- [Myriad::Config](https://metacpan.org/pod/Myriad%3A%3AConfig) - general config support, commandline/file/storage

# METHODS

## loop

Returns the main [IO::Async::Loop](https://metacpan.org/pod/IO%3A%3AAsync%3A%3ALoop) instance for this process.

## services

Hashref of services that have been added to this instance,
as `name` => `Myriad::Service` pairs.

## configure\_from\_argv

Applies configuration from commandline parameters.

Expects a list of parameters and applies the following logic for each one:

- if it contains `::` and a wildcard `*`, it's treated as a service module base name, and all
modules under that immediate namespace will be loaded
- if it contains `::`, it's treated as a comma-separated list of service module names to load
- a `-` prefix is a standard getopt parameter

## transport

Returns the [Myriad::Transport](https://metacpan.org/pod/Myriad%3A%3ATransport) instance according to the config value.

it's designed to be used by tests, so be careful before using it in the framework code.

it takes a single param

- component - the RPC, Subscription or storage in lower case

## redis

The [Net::Async::Redis](https://metacpan.org/pod/Net%3A%3AAsync%3A%3ARedis) (or compatible) instance used for service coördination.

## memory\_transport

The [Myriad::Transport::Memory](https://metacpan.org/pod/Myriad%3A%3ATransport%3A%3AMemory) instance.

## rpc

The [Myriad::RPC](https://metacpan.org/pod/Myriad%3A%3ARPC) instance to serve RPC requests.

## rpc\_client

The [Myriad::RPC::Client](https://metacpan.org/pod/Myriad%3A%3ARPC%3A%3AClient) instance to request other services RPC.

## http

The [Net::Async::HTTP::Server](https://metacpan.org/pod/Net%3A%3AAsync%3A%3AHTTP%3A%3AServer) (or compatible) instance used for health checks
and metrics.

## subscription

The [Myriad::Subscription](https://metacpan.org/pod/Myriad%3A%3ASubscription) instance to manage events.

## storage

The [Myriad::Storage](https://metacpan.org/pod/Myriad%3A%3AStorage) instance to manage data.

## registry

Returns the common [Myriad::Registry](https://metacpan.org/pod/Myriad%3A%3ARegistry) representing the current service state.

## add\_service

Instantiates and adds a new service to the ["loop"](#loop).

Returns the service instance.

## service\_by\_name

Looks up the given service, returning the instance if it exists.

Will throw an exception if the service cannot be found.

## ryu

a source to corresponde to any high level events.

## shutdown

Requests shutdown.

## on\_start

Registers a coderef to be called during startup.
The coderef is expected to return a [Future](https://metacpan.org/pod/Future).

## on\_shutdown

Registers a coderef to be called during shutdown.

The coderef is expected to return a [Future](https://metacpan.org/pod/Future) indicating completion.

## run\_future

Returns a copy of the run [Future](https://metacpan.org/pod/Future).

This would resolve once the process is running and it's
ready to accept requests.

## shutdown\_future

Returns a copy of the shutdown [Future](https://metacpan.org/pod/Future).

This would resolve once the process is about to shut down,
triggered by a fault or a Unix signal.

## setup\_logging

Prepare for logging.

## setup\_tracing

Prepare [OpenTracing](https://metacpan.org/pod/OpenTracing) collection.

## setup\_metrics

Prepare [Metrics::Any::Adapter](https://metacpan.org/pod/Metrics%3A%3AAny%3A%3AAdapter) to collect metrics.

## run

Starts the main loop.

Applies signal handlers for TERM and QUIT, then starts the loop.

# SEE ALSO

Microservices are hardly a new concept, and there's a lot of prior art out there.

Key features that we attempt to provide:

- **reliable handling** - requests and actions should be reliable by default
- **atomic storage** - being able to record something in storage as part of the same transaction as acknowledging a message
- **flexible backends** - support for various storage, RPC and subscription implementations, allowing for mix+match
- **zero transport option** - for testing and smaller deployments, you might want to run everything in a single process
- **language-agnostic** - implementations should be possible in languages other than Perl
- **first-class Kubernetes support** - k8s is not required, but when available we should play to its strengths
- **minimal boilerplate** - with an emphasis on rapid prototyping

These points tend to be incompatible with typical HTTP-based microservices frameworks, although this is
offered as one of the transport mechanisms (with some limitations).

## Perl

Here are a list of the Perl microservice implementations that we're aware of:

- [https://github.com/jmico/beekeeper](https://github.com/jmico/beekeeper) - MQ-based (via STOMP), using [AnyEvent](https://metacpan.org/pod/AnyEvent)
- [https://mojolicious.org](https://mojolicious.org) - more of a web framework, but a popular one
- [Async::Microservice](https://metacpan.org/pod/Async%3A%3AMicroservice) - [AnyEvent](https://metacpan.org/pod/AnyEvent)-based, using HTTP as a protocol, currently a minimal wrapper intended to be used with OpenAPI services

## Java

Although this is the textbook "enterprise-scale platform", Java naturally fits a microservice theme.

- [Spring Boot](https://spring.io/guides/gs/spring-boot/) - One of the frameworks that integrates well
with the traditional Java ecosystem, depends on HTTP as a transport. Although there is no unified storage layer,
database access is available through connectors.
- [Micronaut](https://micronaut.io/) - This framework has many integrations with industry-standard
solutions - SQL, MongoDB, Kafka, Redis, gRPC - and they have integration guides for cloud-native solutions
such as AWS or GCP.
- [DropWizard](https://www.dropwizard.io/en/stable/) - A minimal framework that provides a RESTful
interface and storage layer using Hibernate.
- [Helidon](https://helidon.io/) - Oracle's open source attempt, provides support for two types of
transport and SQL access layer using standard Java's packages, built with cloud-native deployment in mind.

## Python

Most of Python's frameworks provide tools to facilitate building logic blocks behind APIs (Flask, Django ..etc).

For work distribution, [Celery](https://docs.celeryproject.org/en/stable/) is commonly used as a task queue abstraction.

## Rust

- [https://rocket.rs/](https://rocket.rs/) - although this is a web framework, rather than a complete microservice system,
it's reasonably popular for the request/response part of the equation
- [https://actix.rs/](https://actix.rs/) - another web framework, this time with a focus on the actor pattern

## JS

JS has many frameworks that help to implement the microservice architecture, some are:

- [Moleculer](https://moleculer.services/) - generally a full-featured, well-designed microservices framework, highly recommended
- [Seneca](https://senecajs.org/)

## PHP

- [Swoft](http://en.swoft.org/) - async support via Swoole's coroutines, HTTP/websockets based with additional support for Redis/database connection pooling and ORM

## Cloud providers

Microservice support at the provider level:

- [AWS Lambda](https://aws.amazon.com/lambda) - trigger small containers based on logic, typically combined
with other AWS services for data storage, message sending and other actions
- ["Google App Engine"](#google-app-engine) - Google's own attempt
- [Heroku](https://www.heroku.com/) - Allow developers to build a microservices architecture based on the services they provide
like the example they mentioned in this [blog](https://devcenter.heroku.com/articles/event-driven-microservices-with-apache-kafka)

# AUTHOR

Deriv Group Services Ltd. `DERIV@cpan.org`

# CONTRIBUTORS

- Tom Molesworth `TEAM@cpan.org`
- Paul Evans `PEVANS@cpan.org`
- Eyad Arnabeh
- Nael Alolwani

# LICENSE

Copyright Deriv Group Services Ltd 2020-2022. Licensed under the same terms as Perl itself.


Changes

Revision history for {{$dist->name}}

{{$NEXT}}
    [Dependencies]
    - Future::AsyncAwait - updated to v0.61 for some minor bugfixes and Future::XS
    compatibility

    [New features]
    - support for Net::Async::Redis::XS, set the environment variable `PERL_REDIS_XS` to `1`
    to load and use the XS versions for better performance
    - Future::AsyncAwait::Hooks now available (for suspend/resume blocks) when using `:v2`

    [Bugs fixed]
    - handle keyspace notifications properly for config changes in Redis

1.001     2022-11-11 09:42:07+08:00 Asia/Singapore
    [Bugs fixed]
    - Latest Future (and Future::XS) were failing tests due to `Future->needs_all`
    being a class method, the test was incorrectly trying to use it as a plain function
    (only affects the memory transport, not Redis)

1.000     2022-09-05 10:25:25+08:00 Asia/Singapore
    [API changes]
    - The --service_name parameter has been dropped, since it makes no sense when
    used with multiple services or instances

    [New features]
    - Thanks to Syntax::Keyword::Equ, match() now supports `equ` as an operator
    - RPC on start handle existing pending messages in stream, if not expired.
    - RPC overflow protection (will not connsume more than it can process).
    - Transport Redis connection pool count and wait time can be controlled through config parameters.
    - Add orderedset data structure to Myriad storage.

    [Bugs fixed]
    - Subscriptions to create streams from source without a published event.

    [Dependencies]
    - Object::Pad - updated to v0.55 for various bugfixes and improvements

0.010     2021-07-22 07:14:15+00:00 UTC
    [New features]
    - Metrics adapter type is now configurable using --metrics_adapter.
    - Metrics destination can be changed using --metrics_host and --metrics_port.

    [Bugs fixed]
    - RPC client to track message id for response before sending it.
    - Subscription to use the right group name to acknowledge from stream.

    [Dependencies]
    - Object::Pad - use new ->begin_class from MOB package.
    - Metrics::Any - update to v0.07 to disable metrics from certain packages.
    - Metrics::Any::Adapter::Statsd - instead of DogStatsd standard.

0.009     2021-07-14 18:39:34+08:00 Asia/Kuala_Lumpur
    [API Changes]
    - Syntax::Keyword::Match is now imported in Myriad::Service code
    - Redis: A service will not create another service's stream.
    - Allow service namespaces by adding '::*' to the startup command.

    [New features]
    - Added incr method for storage keys.

0.008     2021-06-18 05:24:31+00:00 UTC
    [API Changes]
    - Breaking: RPC client will unwrap the extra 'response' level from the response.
    - Myriad now has run_future like shutdown_future, to tell when the framework is running.
    - Test::Myriad has a ready sub, to tell when the underlying Myriad is running.
    - Batch will now throw an exception and fail if the developer didn't return arrayref.

    [Bugs fixed]
    - Having multiple RPC added will not block and wait each others from reading from stream.

0.007     2021-05-21 14:22:03+00:00 UTC
    [Bugs fixed]
    - RPC reading multiple items from stream but processing only one.

    [Dependencies]
    - Object::Pad - Use officially documented way to obtain metaclasses.

0.006     2021-05-07 04:16:11+00:00 UTC
    [API Changes]
    - `fmap_void`, `fmap_concat` and `fmap_scalar` are imported from Future::Utils into classes by default
    - RPC streams are now organized differently - there will be a stream for each method
    instead of one stream per service. To migrate, use the script in `bin/myriad-migrate-rpc-streams.pl`

    [Bugs fixed]
    - RPC reads no longer not block each other when multiple services have been loaded into one Myriad
    process

0.005     2021-04-30 12:52:41+00:00 UTC
    [Bugs fixed]
    - small memory leak due to defer blocks not being cleared up, now uses try/finally
    to avoid this
    - minor issues in command line parser for services' config
    - bootstrap test to be skipped if Linux::Inotify2 is not available

0.004     2021-04-28 13:27:36+08:00 Asia/Kuala_Lumpur
    [API Changes]
    - Helper methods from List::Util - `min`, `max` and `sum0` - plus JSON functions
    from JSON::MaybeUTF8 now injected into classes by default
    - `defer` keyword from Syntax::Keyword::Defer is also injected into classes by default

    [Bugs fixed]
    - `memory` transport was not delivering messages correctly due to a double-counting bug
    - Redis stream cleanup fixed

0.003     2021-04-27 14:57:49+08:00 Asia/Kuala_Lumpur
    [Bugs fixed]
    - updates to latest Net::Async::Redis to resolve remaining memory leaks
    - numeric configuration values were treated as booleans if the default was set to zero

0.002     2021-04-24 16:36:44+08:00 Asia/Kuala_Lumpur
    [API Changes]
    - The "perl" transport has been renamed to "memory", for consistency
      with implementations in other languages
    - Startup and shutdown calls now receive the `$myriad` instance as the
      first parameter for convenience, and can be method calls (e.g. Myriad.pm
      method name) instead of plain coderefs

    [New features]
    - Subscription to configuration changes via the Redis transport - configuration
      objects still use Ryu::Observable so this behaviour should be a seamless upgrade
      for existing code

    [Bugs fixed]
    - Batch handling had a memory leak, this version should be improved
      although `await` calls inside a batch method may still exhibit
      increased memory usage (we expect to address this in the next version)

0.001     2021-04-16 09:25:26+08:00 Asia/Kuala_Lumpur

 Initial release.


Makefile.PL
# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v6.030.
use strict;
use warnings;



use ExtUtils::MakeMaker 7.64;

my %WriteMakefileArgs = (
  "ABSTRACT" => "async microservice framework",
  "AUTHOR" => "Deriv Services Ltd <DERIV\@cpan.org>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "7.64"
  },
  "DISTNAME" => "Myriad",
  "EXE_FILES" => [
    "bin/myriad-dev.pl",
    "bin/myriad-migrate-rpc-streams.pl",
    "bin/myriad-start.sh",
    "bin/myriad.pl"
  ],
  "LICENSE" => "perl",
  "NAME" => "Myriad",
  "PREREQ_PM" => {
    "Check::UnitCheck" => 0,
    "Class::Method::Modifiers" => 0,
    "Config::Any" => "0.32",
    "Devel::MAT::Dumper" => 0,
    "Future" => "0.50",
    "Future::AsyncAwait" => "0.65",
    "Future::AsyncAwait::Hooks" => "0.02",
    "Future::IO" => "0.13",
    "Future::Queue" => 0,
    "Getopt::Long" => 0,
    "Heap" => "0.80",
    "IO::Async" => "0.802",
    "IO::Async::Notifier" => "0.802",
    "IO::Async::SSL" => "0.24",
    "IO::Async::Test" => "0.802",
    "JSON::MaybeUTF8" => "2.000",
    "List::Keywords" => "0.08",
    "List::Util" => "1.63",
    "Log::Any" => "1.710",
    "Log::Any::Adapter" => "1.710",
    "Log::Any::Adapter::OpenTracing" => "0.001",
    "Math::Random::Secure" => 0,
    "Metrics::Any" => "0.09",
    "Metrics::Any::Adapter::Statsd" => "0.03",
    "Module::Load" => 0,
    "Module::Pluggable::Object" => 0,
    "Module::Runtime" => 0,
    "Net::Async::HTTP" => "0.48",
    "Net::Async::HTTP::Server" => "0.13",
    "Net::Async::OpenTracing" => "1.001",
    "Net::Async::Redis" => "3.022",
    "Object::Pad" => "0.78",
    "OpenTracing::Any" => "1.006",
    "Pod::Simple::Text" => 0,
    "Pod::Usage" => 0,
    "Role::Tiny" => "2.002004",
    "Ryu" => "3.004",
    "Ryu::Async" => "0.020",
    "Scope::Guard" => 0,
    "Syntax::Keyword::Defer" => "0.08",
    "Syntax::Keyword::Dynamically" => "0.12",
    "Syntax::Keyword::Match" => "0.10",
    "Syntax::Keyword::Try" => "0.28",
    "Syntax::Operator::Equ" => "0.05",
    "Sys::Hostname" => 0,
    "Time::Moment" => "0.44",
    "Unicode::UTF8" => 0,
    "XS::Parse::Keyword" => "0.33",
    "XS::Parse::Sublike" => "0.16",
    "YAML::XS" => "0.85",
    "bareword::filehandles" => 0,
    "curry" => "2.000001",
    "indirect" => 0,
    "mro" => 0,
    "multidimensional" => 0
  },
  "TEST_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Log::Any::Adapter::TAP" => 0,
    "Log::Any::Test" => 0,
    "Test::CheckDeps" => "0.010",
    "Test::Deep" => "1.130",
    "Test::Fatal" => "0.014",
    "Test::MemoryGrowth" => "0.03",
    "Test::MockModule" => 0,
    "Test::MockObject" => 0,
    "Test::More" => "0.98",
    "Test::NoTabs" => 0
  },
  "VERSION" => "1.002",
  "test" => {
    "TESTS" => "t/*.t t/RPC/*.t t/Subscription/*.t t/transport/*.t t/util/*.t"
  }
);


my %FallbackPrereqs = (
  "Check::UnitCheck" => 0,
  "Class::Method::Modifiers" => 0,
  "Config::Any" => "0.32",
  "Devel::MAT::Dumper" => 0,
  "ExtUtils::MakeMaker" => 0,
  "File::Spec" => 0,
  "Future" => "0.50",
  "Future::AsyncAwait" => "0.65",
  "Future::AsyncAwait::Hooks" => "0.02",
  "Future::IO" => "0.13",
  "Future::Queue" => 0,
  "Getopt::Long" => 0,
  "Heap" => "0.80",
  "IO::Async" => "0.802",
  "IO::Async::Notifier" => "0.802",
  "IO::Async::SSL" => "0.24",
  "IO::Async::Test" => "0.802",
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "JSON::MaybeUTF8" => "2.000",
  "List::Keywords" => "0.08",
  "List::Util" => "1.63",
  "Log::Any" => "1.710",
  "Log::Any::Adapter" => "1.710",
  "Log::Any::Adapter::OpenTracing" => "0.001",
  "Log::Any::Adapter::TAP" => 0,
  "Log::Any::Test" => 0,
  "Math::Random::Secure" => 0,
  "Metrics::Any" => "0.09",
  "Metrics::Any::Adapter::Statsd" => "0.03",
  "Module::Load" => 0,
  "Module::Pluggable::Object" => 0,
  "Module::Runtime" => 0,
  "Net::Async::HTTP" => "0.48",
  "Net::Async::HTTP::Server" => "0.13",
  "Net::Async::OpenTracing" => "1.001",
  "Net::Async::Redis" => "3.022",
  "Object::Pad" => "0.78",
  "OpenTracing::Any" => "1.006",
  "Pod::Simple::Text" => 0,
  "Pod::Usage" => 0,
  "Role::Tiny" => "2.002004",
  "Ryu" => "3.004",
  "Ryu::Async" => "0.020",
  "Scope::Guard" => 0,
  "Syntax::Keyword::Defer" => "0.08",
  "Syntax::Keyword::Dynamically" => "0.12",
  "Syntax::Keyword::Match" => "0.10",
  "Syntax::Keyword::Try" => "0.28",
  "Syntax::Operator::Equ" => "0.05",
  "Sys::Hostname" => 0,
  "Test::CheckDeps" => "0.010",
  "Test::Deep" => "1.130",
  "Test::Fatal" => "0.014",
  "Test::MemoryGrowth" => "0.03",
  "Test::MockModule" => 0,
  "Test::MockObject" => 0,
  "Test::More" => "0.98",
  "Test::NoTabs" => 0,
  "Time::Moment" => "0.44",
  "Unicode::UTF8" => 0,
  "XS::Parse::Keyword" => "0.33",
  "XS::Parse::Sublike" => "0.16",
  "YAML::XS" => "0.85",
  "bareword::filehandles" => 0,
  "curry" => "2.000001",
  "indirect" => 0,
  "mro" => 0,
  "multidimensional" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);
